server/src/utils/Utils.ts

```diff
@@ -21,6 +25,44 @@ export const selectRandomMany = (items, n) => {
   return _.times(n, () => items[Math.floor(Math.random() * items.length)]);
 };
 
+export const selectRandomWeightedNoReplacement = (items, weights, n) => {
+  if (items.length !== weights.length) throw new Error("Items and weights must be same length!");
+
+  // If weights are all 0, set all values to 1
+  // Otherwise, array of undefined elements is returned
+  if (!weights.some((weight) => weight !== 0)) weights = weights.map((weight) => weight + 1);
+
+  let sum = 0;
+  let segments = new Array();
+  for (let i = 0; i < weights.length; i++) {
+    sum += weights[i];
+    segments.push(sum);
+  }
+
+  let choices = new Array();
+  for (let i = 0; i < n; i++) {
+    const random = Math.random() * sum;
+    const index = segments.filter((segment) => segment <= random).length;
+
+    const item = items[index];
+    const newItem = _.cloneDeep(item);
+    choices.push(newItem);
+
+    // remove & reweight
+    items.splice(index, 1);
+    weights.splice(index, 1);
+    if (weights.length > 0 && items.length > 0) {
+      segments = new Array();
+      sum = weights.reduce((accum, curr) => {
```

Comment:
This was some late night hackathon energy I apologize. Its def performant enough for the # of elements we have though.

Reply:
lol no worries, as long as it works.